Draw an outline for 2D debug collision shapes

Calinou opened the pull request:
This makes them easier to distinguish, especially when used in a TileMap or when they're placed next to each other.

The default color's opacity has been slightly decreased to account for the new outline.

## Preview

### Before

![visible_collision_shapes_before](https://user-images.githubusercontent.com/180032/72000959-ad9af500-3244-11ea-8782-dac486cacdee.png)

### After

![visible_collision_shapes_after](https://user-images.githubusercontent.com/180032/72000957-ac69c800-3244-11ea-8b8c-e6b9682e60ca.png)

Xrayez commented:
Does `draw_polyline` draws the closed polylines here? I typically get this result when attempting to draw a "non-filled" polygon, say this :

```gdscript
draw_polyline(PoolVector2Array([Vector2(0,0), Vector2(100, 0), Vector2(100, 100), Vector2(0, 100)]), Color.green)
```
![godot-polyline-open](https://user-images.githubusercontent.com/17108460/72001846-eb048000-324e-11ea-8f72-a5145bd1a707.png)

And then I just draw the last segment manually with `draw_line`:
```gdscript
draw_line(Vector2(0,0), Vector2(0, 100), Color.green)
```

I wish this was present in API but I might be missing something.

Calinou commented:
@Xrayez It seems it doesn't add the last point automatically, so a missing line can be noticed on large CapsuleShapes/CircleShapes (or when zooming in). I'll look into fixing it.

Xrayez commented on scene/2d/collision_polygon_2d.cpp:
```diff
@@ -159,6 +159,7 @@ void CollisionPolygon2D::_notification(int p_what) {
 			}
 #else
 			draw_colored_polygon(polygon, get_tree()->get_debug_collisions_color());
+			draw_polyline(polygon, get_tree()->get_debug_collisions_color());
```
Missed `draw_line` here as well?

Calinou replied:
Is this part used at all? I noticed CollisionPolygon2Ds appear in a different color and already have an outline anyway.

Xrayez replied:
If `DEBUG_DECOMPOSE` is configurable then it might make sense, else this could be handled by `ProjectSettings` instead?

pouleyKetchoupp replied:
This code is used in non-tool debug to display a unique shape instead of the decomposed convex shapes.

Adding `draw_polyline` and `draw_line` here would draw an extra line on top of the red line in this case, so this block shouldn't change.

avencherus commented:
@Xrayez Noticed this too, it seems to close fine if the starting point is also appended to the end.

`	draw_polyline(PoolVector2Array([Vector2(0,0), Vector2(100, 0), Vector2(100, 100), Vector2(0, 100), Vector2(0,0)]), Color.green)`

Xrayez commented:
Yeah, yet the issue reveals itself more when you're trying to draw a polygon with lines, and when you don't have an opportunity to append another point. The method could benefit from `closed` param in that case, but that's proposal material.

eon-s commented:
I guess this and other of the great visual improvements made by Calinou will end on 4.0, but could all these PRs be added to cherry pick later for 3.2.1?

aaronfranke commented:
@Calinou Any update? This needs to be rebased on the latest master branch.

I didn't know that I needed this, but now that I see this PR, I want this a lot :)

Calinou committed:
Draw an outline for 2D debug collision shapes

This makes them easier to distinguish, especially when used
in a TileMap.

The default color's opacity has been slightly decreased to account
for the new outline.

Calinou commented:
I remade this pull request from scratch following the physics changes.

KoBeWi commented on scene/resources/rectangle_shape_2d.cpp:
```diff
@@ -47,11 +47,27 @@ Vector2 RectangleShape2D::get_size() const {
 
 void RectangleShape2D::draw(const RID &p_to_rid, const Color &p_color) {
-	RenderingServer::get_singleton()->canvas_item_add_rect(p_to_rid, Rect2(-size / 2, size), p_color);
+	// Draw an outlined rectangle to make individual shapes easier to distinguish.
+	Vector<Vector2> stroke_points;
+	stroke_points.resize(5);
+	stroke_points.write[0] = -size * 0.5;
+	stroke_points.write[1] = Vector2(size.x, -size.y) * 0.5;
+	stroke_points.write[2] = size * 0.5;
+	stroke_points.write[3] = Vector2(-size.x, size.y) * 0.5;
+	stroke_points.write[4] = -size * 0.5;
+
+	Vector<Color> stroke_colors;
+	stroke_colors.resize(5);
+	for (int i = 0; i < 5; i++) {
+		stroke_colors.write[i] = (p_color);
+	}
+
+	RenderingServer::get_singleton()->canvas_item_add_rect(p_to_rid, Rect2(-size * 0.5, size), p_color);
+	RenderingServer::get_singleton()->canvas_item_add_polyline(p_to_rid, stroke_points, stroke_colors);
```
I think this can be replaced simply by `draw_rect()`.

Calinou replied:
@KoBeWi There's no `draw_rect()` in RenderingServer.

KoBeWi replied:
Ah nevermind, I think I confused it with CollisionShape2D.

pouleyKetchoupp reviewed (changes requested):
That looks great!

For the `CollisionPolygon2D` case, apart from the change I've mentioned, I'd like to actually refactor it a little bit since you're updating collision debug draw.

It doesn't make a lot of sense for polygon shapes to be displayed in a different color as other shapes. So here's my suggestion to make things more consistent:
```
			Color color = get_tree()->get_debug_collisions_color();

			draw_polyline(polygon, color);
			// Draw the last segment as it's not drawn by `canvas_item_add_polyline()`.
			draw_line(polygon[polygon.size() - 1], polygon[0], color);

#define DEBUG_DECOMPOSE
#if defined(TOOLS_ENABLED) && defined(DEBUG_DECOMPOSE)
			Vector<Vector<Vector2> > decomp = _decompose_in_convex();

			Color c = color;
			float min_value = 0.2;
			for (int i = 0; i < decomp.size(); i++) {
				draw_colored_polygon(decomp[i], c);
				float new_value = Math::fmod(c.get_v() + 0.738, 1.0);
				if (new_value < min_value) {
					new_value += min_value;
				}
				c.set_hsv(c.get_h(), c.get_s(), new_value, c.a);
			}
#else
			draw_colored_polygon(polygon, color);
#endif
```

It also alternates values instead of hues when drawing the different convex shapes from decomposition, which is colorblind-friendly and easier to read because it matches the style better.

Before:
<img width="400" src="https://user-images.githubusercontent.com/1075032/108545308-dfceb100-72a4-11eb-9595-ec99e3889807.png">

After:
<img width="400" src="https://user-images.githubusercontent.com/1075032/108545371-efe69080-72a4-11eb-8b9c-2bd5509b5d4b.png">

Would that make sense to you? I can also make it a separate PR in case it's too much out of scope for this PR.

Calinou commented:
> Would that make sense to you? I can also make it a separate PR in case it's too much out of scope for this PR.

Sure, but I think this should be done in a separate PR :slightly_smiling_face:

akien-mga commented:
Thanks!

akien-mga commented:
This would need a dedicated PR for `3.2` as it's not trivial to cherry-pick.
